lib/manager/pip-compile/artifacts.ts

```diff
@@ -0,0 +1,102 @@
+import { quote as pipCompile } from 'shlex';
+import { TEMPORARY_ERROR } from '../../constants/error-messages';
+import { logger } from '../../logger';
+import { ExecOptions, exec } from '../../util/exec';
+import { deleteLocalFile, readLocalFile, writeLocalFile } from '../../util/fs';
+import { getRepoStatus } from '../../util/git';
+import type {
+  UpdateArtifact,
+  UpdateArtifactsConfig,
+  UpdateArtifactsResult,
+} from '../types';
+
+function getPythonConstraint(
+  config: UpdateArtifactsConfig
+): string | undefined | null {
+  const { constraints = {} } = config;
+  const { python } = constraints;
+
+  if (python) {
+    logger.debug('Using python constraint from config');
+    return python;
+  }
+
+  return undefined;
+}
+
+function getPipToolsConstraint(config: UpdateArtifactsConfig): string | null {
+  const { constraints = {} } = config;
+  const { pipTools } = constraints;
+
+  if (pipTools) {
+    logger.debug('Using pipTools constraint from config');
+    return pipTools;
+  }
+
+  return '';
+}
+
+export async function updateArtifacts({
+  packageFileName: inputFileName,
+  newPackageFileContent: newInputContent,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  const outputFileName = inputFileName.replace(/(\.in)?$/, '.txt');
+  logger.debug(
+    `pipCompile.updateArtifacts(${inputFileName}->${outputFileName})`
+  );
+  const existingOutput = await readLocalFile(outputFileName, 'utf8');
+  if (!existingOutput) {
+    logger.debug('No pip-compile output file found');
+    return null;
+  }
+  try {
+    await writeLocalFile(inputFileName, newInputContent);
+    if (config.isLockFileMaintenance) {
+      await deleteLocalFile(outputFileName);
+    }
+    const cmd = 'pip-compile';
+    const tagConstraint = getPythonConstraint(config);
+    const pipToolsConstraint = getPipToolsConstraint(config);
+    const execOptions: ExecOptions = {
+      cwdFile: inputFileName,
+      docker: {
+        image: 'python',
+        tagConstraint,
+        tagScheme: 'pep440',
+        preCommands: [
+          `pip install --user ${pipCompile(`pip-tools${pipToolsConstraint}`)}`,
```

Comment:
What about buildpack `install-pip` command? 🤔

Reply:
Let's refactor that at the same time as pipenv etc. They all or mostly use `pip install`